addons/gamification/models/res_users.py

```diff
@@ -96,6 +112,66 @@ def _recompute_rank(self):
             if old_rank != user.rank_id:
                 user._rank_changed()
 
+    def _recompute_rank_bulk(self):
+        """
+            Compute rank of each user by rank.
+            For each rank, check which users need to be ranked
+
+        """
+        ranks = [{'rank': rank, 'karma_min': rank.karma_min} for rank in
+                 self.env['gamification.karma.rank'].search([], order="karma_min DESC")]
+
+        users_todo = self
+
+        next_rank_id = False
+        # wtf, next_rank_id should be a related on rank_id.next_rank_id and life might get easier.
+        # And we only need to recompute next_rank_id on write with min_karma or in the create on rank model.
+        for r in ranks:
+            rank_id = r['rank'].id
+            dom = [
+                ('karma', '>=', r['karma_min']),
+                ('id', 'in', users_todo.ids),
+                '|',  # noqa
+                    '|', ('rank_id', '!=', rank_id), ('rank_id', '=', False),
+                    '|', ('next_rank_id', '!=', next_rank_id), ('next_rank_id', '=', False if next_rank_id else -1),
```

Comment:
Shouldn't that be better to process the last condition in python? This would avoid a slow computation in postgres

```python
next_rank_condition = [('next_rank_id', '=', False)] if next_rank_id else []

dom = [
    ('karma', '>=', r['karma_min']),
    ('id', 'in', users_todo.ids),
    '|',  # noqa
        '|', ('rank_id', '!=', rank_id), ('rank_id', '=', False),
        expression.OR([[('next_rank_id', '!=', next_rank_id)], next_rank_condition]),
]
```

Reply:
tu parles d'un  slow computation in postgres de qque ms ? 
car de toute facon il doit chercher sur next_rank_id dans la premier condition.

Donc pas convaincu du gain de performance, mais si tu preferes, on peut changer ca ;)

Sinon en pratique on passe de qques minutes/heures -> à qques secondes... mais on peut passer a (qques secondes - qques milisecondes) si tu preferes :+1:

Reply:
not so easy since expression.OR will return as well [('a','=','b')] as well ['|', ('a','=','b'), ('c','=','d')] so you need to flat your new array and co.

Since this code need a refacoring in master and so before v13, we merge like that for now.